src/serac/physics/state/state_manager.cpp

```diff
@@ -20,139 +20,166 @@ QuadratureData<void> dummy_qdata;
+    FiniteElementDual dual(mesh(tag), std::move(options));
 
     // Create a grid function view of the local vector for plotting
     // Note: this is a static cast because we know this vector under the hood is a grid function
     // This is hidden from the user because the interpolation capabilities of a grid function
     // are inappropriate for dual vectors.
     auto gf_view_of_local_dual_vector = static_cast<mfem::ParGridFunction*>(&dual.localVec());
 
-    datacoll_->RegisterField(name, gf_view_of_local_dual_vector);
+    datacoll.RegisterField(name, gf_view_of_local_dual_vector);
     // Now that it's been allocated, we can set it to zero
     dual.localVec() = 0.0;
     return dual;
   }
 }
 
-void StateManager::save(const double t, const int cycle)
+void StateManager::save(const double t, const int cycle, const std::string& tag)
 {
-  SLIC_ERROR_ROOT_IF(!datacoll_, "Serac's datacollection was not initialized - call StateManager::initialize first");
-
-  std::string file_path =
-      axom::utilities::filesystem::joinPath(datacoll_->GetPrefixPath(), datacoll_->GetCollectionName());
+  SLIC_ERROR_ROOT_IF(!ds_, "Serac's datacollection was not initialized - call StateManager::initialize first");
+  auto&       datacoll  = datacolls_.at(tag);
+  std::string file_path = axom::utilities::filesystem::joinPath(datacoll.GetPrefixPath(), datacoll.GetCollectionName());
   SLIC_INFO_ROOT(axom::fmt::format("Saving data collection at time: {} to path: {}", t, file_path));
 
   for (const auto& data : syncable_data_) {
     data->sync();
   }
-  datacoll_->SetTime(t);
-  datacoll_->SetCycle(cycle);
-  datacoll_->Save();
+  datacoll.SetTime(t);
+  datacoll.SetCycle(cycle);
+  datacoll.Save();
 }
 
-void StateManager::setMesh(std::unique_ptr<mfem::ParMesh> pmesh)
+mfem::ParMesh* StateManager::setMesh(std::unique_ptr<mfem::ParMesh> pmesh, const std::string& tag)
 {
-  datacoll_->SetMesh(pmesh.release());
-  datacoll_->SetOwnData(true);
+  newDataCollection(tag);
+  auto& datacoll = datacolls_.at(tag);
+  datacoll.SetMesh(pmesh.release());
+  datacoll.SetOwnData(true);
 
   // Functional needs the nodal grid function and neighbor data in the mesh
-  StateManager::mesh().EnsureNodes();
-  StateManager::mesh().ExchangeFaceNbrData();
+  auto& new_pmesh = mesh(tag);
+  new_pmesh.EnsureNodes();
+  new_pmesh.ExchangeFaceNbrData();
+  return &new_pmesh;
 }
 
-mfem::ParMesh& StateManager::mesh()
+mfem::ParMesh& StateManager::mesh(const std::string& tag)
 {
-  auto mesh = datacoll_->GetMesh();
-  SLIC_ERROR_ROOT_IF(!mesh, "The datastore does not contain a mesh object");
+  auto mesh = datacolls_.at(tag).GetMesh();
+  SLIC_ERROR_ROOT_IF(!mesh, "The datacollection does not contain a mesh object");
   return static_cast<mfem::ParMesh&>(*mesh);
 }
 
+std::string StateManager::collectionID(mfem::ParMesh* pmesh)
+{
+  if (!pmesh) {
+    return default_mesh_name_;
+  } else {
+    for (auto& [name, datacoll] : datacolls_) {
+      if (datacoll.GetMesh() == pmesh) {
```

Comment:
This pointer comparison makes me a little uneasy. Is it ever possible they could be moved?

Reply:
The mesh pointer owned by the datacollection does not change, if that's what you're asking

Reply:
As long as the raw pointer comparison is documented, I'm OK with this.